sw/device/tests/spi_device_sleep_test.c

```diff
@@ -0,0 +1,157 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_aon_timer.h"
+#include "sw/device/lib/dif/dif_pinmux.h"
+#include "sw/device/lib/dif/dif_pwrmgr.h"
+#include "sw/device/lib/dif/dif_rv_core_ibex.h"
+#include "sw/device/lib/dif/dif_rv_plic.h"
+#include "sw/device/lib/runtime/irq.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/aon_timer_testutils.h"
+#include "sw/device/lib/testing/pwrmgr_testutils.h"
+#include "sw/device/lib/testing/rv_plic_testutils.h"
+#include "sw/device/lib/testing/spi_device_testutils.h"
+#include "sw/device/lib/testing/spi_flash_emulator.h"
+#include "sw/device/lib/testing/spi_flash_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "sw/device/lib/testing/autogen/isr_testutils.h"
+
+OTTF_DEFINE_TEST_CONFIG();
+
+enum {
+  kWakeupTimeMicros = 500000,
+};
+static dif_spi_device_handle_t spid;
+static dif_pwrmgr_t pwrmgr;
+static dif_aon_timer_t aon_timer;
+static dif_rv_plic_t plic;
+static plic_isr_ctx_t plic_ctx = {.rv_plic = &plic,
+                                  .hart_id = kTopEarlgreyPlicTargetIbex0};
+static pwrmgr_isr_ctx_t pwrmgr_isr_ctx = {
+    .pwrmgr = &pwrmgr,
+    .plic_pwrmgr_start_irq_id = kTopEarlgreyPlicIrqIdPwrmgrAonWakeup,
+    .expected_irq = kDifPwrmgrIrqWakeup,
+    .is_only_irq = true};
+
+static void enter_low_power(void) {
+  uint64_t wkup_cycles = 0;
+  CHECK_STATUS_OK(aon_timer_testutils_get_aon_cycles_64_from_us(
+      kWakeupTimeMicros, &wkup_cycles));
+  CHECK_STATUS_OK(aon_timer_testutils_wakeup_config(&aon_timer, wkup_cycles));
+
+  dif_pwrmgr_domain_config_t pwrmgr_domain_cfg =
+      kDifPwrmgrDomainOptionMainPowerInLowPower;
+
+  LOG_INFO("SYNC: Entering lowpower mode");
+  CHECK_STATUS_OK(pwrmgr_testutils_enable_low_power(
+      &pwrmgr, kDifPwrmgrWakeupRequestSourceFive, pwrmgr_domain_cfg));
+  wait_for_interrupt();
```

Comment:
There's a bit of a race with the host here, since low power mode hasn't been enabled yet. It takes dozens of microseconds to set up the transition to low power mode, and the host could have processed that UART message and issued a SPI transaction before we ever reach WFI.

Do we need to have completed the transition to low power for this test? Or is it okay that we might be hitting spi_device while pwrmgr is still transitioning? The pinmux will take on the sleep values almost immediately after pwrmgr recognizes the WFI, so testing outputs during the transition should be fine, I think.

Reply:
I think it's ok for this test to issue the transaction before the core is sleeping. But if I want to test with core in sleep, what would be the solution? A delay on the host side?

Reply:
You could make use of a GPIO indicator. You'd have some default value indicating the core is active, then use a deep sleep function in the pinmux to drive it the other direction.

For example, the host could pull up on a high-z MIO, then the pinmux's deep sleep mode for that MIO could be set to Tie-Low. Then the host can poll that MIO and wait to see 0.

Reply:
Would that also work for normal sleep?

Reply:
> Would that also work for normal sleep?

Oh, judging by the connectivity and RTL, yes. I think I must've somehow connected the pinmux CSRs with deep sleep, but that doesn't appear to be quite right. They are just "sleep modes" and pwrmgr gives the same "low_power_o" output for normal sleep and deep sleep.

Reply:
Done. I'm waking up the core with a gpio and checking a gpio to make sure it entered low power mode as suggested.